### DISCUSSION TITLE
Embedded j9stubs.jar ?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-equinox/equinox#131] [Q&A] [answered]
Similar to the[ last question](https://github.com/eclipse-equinox/equinox/discussions/130), there is a `j9stubs.jar `embedded in the equinox, what is it used for? Is equinox actually supposed to be some kind of [Multi-Release jar](https://openjdk.org/jeps/238)? Is it to prepare for move to java 9+?

**Comment by tjwatson** [comment, marked as answer]
This is so we can compile in support for shared classes cache when running on Eclipse OpenJ9 Java.  We need the `com.ibm.oti.shared` package to do that.  See `org.eclipse.osgi.internal.cds` package.  Also see https://blog.openj9.org/2018/10/30/openj9-shared-classes-with-eclipse-equinox-4-9/

This jar is here so we can continue to compile with hotspot JDK and not require OpenJ9 JDK to compile Equinox.

**Comment by tjwatson** [comment]
With that said, we could change to using reflection like we do in OpenLiberty: https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.kernel.boot.common/src/com/ibm/ws/kernel/boot/classloader/SharedClassCacheHook.java

But I find non-reflection to be much better to understand.

**Comment by laeubi** [comment]
Thanks for the insights, this seems not so easy to get rid of ...

**Comment by laeubi** [comment]
@tjwatson one thing that came into my mind. Do you think it would be possible that what currently is `j9stubs.jar` to be released by the J9 project as a regular maven-artifact?

That way it would be possible to reference that in a maven standard way as a `provided` dependency without the need to embed the jar, other projects can reuse it and we won't need tycho/pde specific `jars.extra.classpath`?

**Reply by tjwatson** [reply]
We could ask, but it would be a little strange in my opinion because I doubt the artifact would be anything that could be useful at runtime.  It would have to be a stubbed set of APIs they have for shared classes that is only meant to compile against.

**Reply by laeubi** [reply]
> I doubt the artifact would be anything that could be useful at runtime.

It won't be used at runtime or installed anywhere (`provided` in maven means = use for compile but it is not a regular dependency but a perquisite of the platform where the artifact runs), similar to the osgi-annotation jars.

>  It would have to be a stubbed set of APIs they have for shared classes that is only meant to compile against.

Yes exactly that would be the idea, so even OpenLiberty could use this as a `provided` dependency (or others as well) but dont need to build/host an own stubbed API...

**Reply by tjwatson** [reply]
We can ask, I'm just not sure they will be up for doing it.

**Reply by laeubi** [reply]
Can you bring this up? I have absolute no knowledge about J9 and how it works or which people best to ask here and what are the exact demands of Equinox why we need this...

**Reply by tjwatson** [reply]
I will, not sure on the chances of success.  Another option is to compile the stub source as part of the `org.eclipse.osg` project that we can then compile against but not ship?
